config/src/main/java/com/typesafe/config/impl/Tokens.java

```diff
@@ -222,7 +240,8 @@ public int hashCode() {
         final private List<Token> value;
 
         Substitution(ConfigOrigin origin, boolean optional, List<Token> expression) {
-            super(TokenType.SUBSTITUTION, origin);
+            super(TokenType.SUBSTITUTION, origin,
+                    "${" + (optional? "?" : "") + Tokenizer.render(expression.iterator()) + "}");
```

Comment:
this makes me wonder if tokenText() should be virtual instead of final, so we don't need to precompute and keep around a field here that's completely derivable from the other fields we already have.